include/gunrock/algorithms/tc.hxx

```diff
@@ -0,0 +1,152 @@
+/**
+ * @file tc.hxx
+ * @author Muhammad A. Awad (mawad@ucdavis.edu)
+ * @brief Triangle Counting algorithm.
+ * @version 0.1
+ * @date 2022-08-06
+ *
+ * @copyright Copyright (c) 2022
+ *
+ */
+#pragma once
+
+#include <gunrock/algorithms/algorithms.hxx>
+
+namespace gunrock {
+namespace tc {
+
+template <typename vertex_t>
+struct param_t {
+  bool reduce_all_triangles;
+  param_t(bool _reduce_all_triangles)
+      : reduce_all_triangles(_reduce_all_triangles) {}
+};
+
+template <typename vertex_t>
+struct result_t {
+  vertex_t* vertex_triangles_count;
+  std::size_t* total_triangles_count;
+  result_t(vertex_t* _vertex_triangles_count, uint64_t* _total_triangles_count)
+      : vertex_triangles_count(_vertex_triangles_count),
+        total_triangles_count(_total_triangles_count) {}
+};
+
+template <typename graph_t, typename param_type, typename result_type>
+struct problem_t : gunrock::problem_t<graph_t> {
+  param_type param;
+  result_type result;
+
+  problem_t(graph_t& G,
+            param_type& _param,
+            result_type& _result,
+            std::shared_ptr<gcuda::multi_context_t> _context)
+      : gunrock::problem_t<graph_t>(G, _context),
+        param(_param),
+        result(_result) {}
+
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t = typename graph_t::edge_type;
+
+  void init() override {}
+
+  void reset() override {}
+};
+
+template <typename problem_t>
+struct enactor_t : gunrock::enactor_t<problem_t> {
+  enactor_t(problem_t* _problem,
+            std::shared_ptr<gcuda::multi_context_t> _context)
+      : gunrock::enactor_t<problem_t>(_problem, _context) {}
+
+  using vertex_t = typename problem_t::vertex_t;
+  using edge_t = typename problem_t::edge_t;
+  using weight_t = typename problem_t::weight_t;
+  using frontier_t = typename enactor_t<problem_t>::frontier_t;
+
+  void prepare_frontier(frontier_t* f,
+                        gcuda::multi_context_t& context) override {
+    auto P = this->get_problem();
+    auto n_vertices = P->get_graph().get_number_of_vertices();
+
+    f->sequence((vertex_t)0, n_vertices, context.get_context(0)->stream());
```

Comment:
If you're queuing all vertices, i.e. advancing over the whole graph, it may be worth using the correct i/o for the advance instead of a frontier of all vertices. See the comment below for the other method.